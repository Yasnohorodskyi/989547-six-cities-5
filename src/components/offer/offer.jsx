import React from "react";
import CommentForm from "../comment-form/comment-form";
import propTypes from "prop-types";
import ReviewsList from "../reviews-list/reviews-list.jsx";
import Map from "../map/map";
import NearOffersList from "../near-offers-list/near-offers-list";
import {withDataLoading} from "../../hocks/with-data-loading";
import {connect} from "react-redux";
import classNames from "classnames";
import {Link} from "react-router-dom";
import UserNav from "../user-nav/user-nav";
import {AuthorizationStatus} from "../../consts";
import BookmarkButton from "../bookmarkButton/bookmarkButton";
import {getComments, getNearbyOffers, getOfferById} from "../../store/api-action";

const Offer = (props) => {
  const {offer, comments, nearbyHotels, authorizationStatus} = props;
  const {isPremium, description, bedrooms, adults, cost, isFavorite, rate, title, type, host, id} = offer;
  const {avatarUrl, name, isPro} = host;

  const images = offer.images.map((image, index) => ({
    url: image,
    id: index
  }));

  const goods = offer.goods.map((good, index) => ({
    name: good,
    id: index
  }));

  const commentComponent = () => {
    if (authorizationStatus === AuthorizationStatus.AUTH) {
      return <CommentForm id={id}/>;
    }
    return null;
  };


  return (
    <div className="page">
      <header className="header">
        <div className="container">
          <div className="header__wrapper">
            <div className="header__left">
              <Link className="header__logo-link" to="/">
                <img className="header__logo" src="img/logo.svg" alt="6 cities logo" width="81" height="41" />
              </Link>
            </div>
            <UserNav />
          </div>
        </div>
      </header>

      <main className="page__main page__main--property">
        <section className="property">
          <div className="property__gallery-container container">
            <div className="property__gallery">
              {images.map((image) => {
                return (<div key={image.id} className="property__image-wrapper">
                  <img className="property__image" src={image.url} alt="Photo studio" />
                </div>);
              })}
            </div>
          </div>
          <div className="property__container container">
            <div className="property__wrapper">
              {isPremium
                ? <div className="property__mark">
                  <span>Premium</span>
                </div>
                : ``}
              <div className="property__name-wrapper">
                <h1 className="property__name">
                  {title}
                </h1>
                <BookmarkButton name="property" id={id} isFavorite={isFavorite} width={31} height={33}/>
              </div>
              <div className="property__rating rating">
                <div className="property__stars rating__stars">
                  <span style={{width: `${rate * 20}%`}}></span>
                  <span className="visually-hidden">Rating</span>
                </div>
                <span className="property__rating-value rating__value">{rate}</span>
              </div>
              <ul className="property__features">
                <li className="property__feature property__feature--entire">
                  {type.charAt(0).toUpperCase() + type.slice(1)}
                </li>
                <li className="property__feature property__feature--bedrooms">
                  {bedrooms} Bedrooms
                </li>
                <li className="property__feature property__feature--adults">
                  Max {adults} adults
                </li>
              </ul>
              <div className="property__price">
                <b className="property__price-value">&euro;{cost}</b>
                <span className="property__price-text">&nbsp;night</span>
              </div>
              <div className="property__inside">
                <h2 className="property__inside-title">What&apos;s inside</h2>
                <ul className="property__inside-list">
                  {goods.map((good) => {
                    return (
                      <li key={good.id} className="property__inside-item">
                        {good.name}
                      </li>
                    );
                  })}
                </ul>
              </div>
              <div className="property__host">
                <h2 className="property__host-title">Meet the host</h2>
                <div className="property__host-user user">
                  <div className={classNames(`property__avatar-wrapper user__avatar-wrapper`, {"property__avatar-wrapper--pro": isPro})}>
                    <img className="property__avatar user__avatar" src={avatarUrl} width="74" height="74" alt="Host avatar" />
                  </div>
                  <span className="property__user-name">
                    {name}
                  </span>
                </div>
                <div className="property__description">
                  <p className="property__text">
                    {description}
                  </p>
                </div>
              </div>
              <section className="property__reviews reviews">
                <h2 className="reviews__title">Reviews &middot; <span className="reviews__amount">{comments.length}</span></h2>

                <ReviewsList comments={comments}/>
                {commentComponent()}
              </section>
            </div>
          </div>
          <section className="property__map map">
            <Map offers={nearbyHotels} styles={{width: `1144px`, marginLeft: `auto`, marginRight: `auto`}}/>
          </section>
        </section>
        <div className="container">
          <section className="near-places places">
            <h2 className="near-places__title">Other places in the neighbourhood</h2>
            <NearOffersList offers={nearbyHotels} />
          </section>
        </div>
      </main>
    </div>
  );
};

Offer.propTypes = {
  offer: propTypes.object.isRequired,
  comments: propTypes.array.isRequired,
  nearbyHotels: propTypes.array.isRequired,
  authorizationStatus: propTypes.string.isRequired
};

const mapStateToProps = ({PROCESS, USER}) => ({
  offer: PROCESS.activeOffer,
  comments: PROCESS.comments,
  nearbyHotels: PROCESS.nearbyHotels,
  authorizationStatus: USER.authorizationStatus,
  isDataLoaded: PROCESS.isOfferLoaded && PROCESS.isCommentsLoaded && PROCESS.isNearbyOffersLoaded,
});

const mapDispatchToProps = (dispatch) => ({
  loadComments(id) {
    dispatch(getComments(id));
  },
  loadOffer(id) {
    dispatch(getOfferById(id));
  },
  loadNearbyOffers(id) {
    dispatch(getNearbyOffers(id));
  },
});

export {Offer};
export default connect(mapStateToProps, mapDispatchToProps)(withDataLoading(Offer));
