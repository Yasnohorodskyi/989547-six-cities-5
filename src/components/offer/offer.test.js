import React from "react";
import renderer from "react-test-renderer";
import {Provider} from "react-redux";
import {MemoryRouter} from "react-router-dom";
import configureMockStore from "redux-mock-store";

import {Offer} from "./offer";

const offer = {
  adults: 0,
  bedrooms: 0,
  city: {
    name: ``,
  },
  cost: 0,
  description: `Hello, mello, shmello`,
  goods: [],
  host: {
    id: 0,
    name: ``,
    avatarUrl: ``,
    isPro: true
  },
  id: 0,
  images: [],
  isFavorite: true,
  isPremium: true,
  previewImg: ``,
  rate: 0,
  title: ``,
  type: ``
};

const mockStore = configureMockStore()({
  PROCESS: {
    offer,
    comments: [],
    nearbyHotels: [],
  },
  USER: {
    authorizationStatus: ``,
    isDataLoaded: true
  }
});

const authorizationStatus = `AUTH`;
const comments = [
  {
    comment: ``,
    date: `2020-12-02T17:54:54.499Z`,
    id: 0,
    rating: 5,
    user: {
      id: 1,
      name: ``,
      avatarUrl: ``,
      isPro: true,
    }
  }
];
const nearbyHotels = [
  offer
];

jest.mock(`../city-map/city-map`, () => `CityMap`);
jest.mock(`../user-nav/user-nav`, () => `UserNav`);
jest.mock(`../comment-form/comment-form`, () => `CommentForm`);

it(`Should offer page render`, () => {
  const tree = renderer
  .create(
      <Provider store={mockStore}>
        <MemoryRouter>
          <Offer
            offer={offer}
            authorizationStatus={authorizationStatus}
            comments={comments}
            nearbyHotels={nearbyHotels}
          />
        </MemoryRouter>
      </Provider>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
