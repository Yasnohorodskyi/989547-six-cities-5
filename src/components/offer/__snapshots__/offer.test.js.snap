// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should offer page render 1`] = `
<div
  className="page"
>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <div
        className="header__wrapper"
      >
        <div
          className="header__left"
        >
          <a
            className="header__logo-link"
            href="/"
            onClick={[Function]}
          >
            <img
              alt="6 cities logo"
              className="header__logo"
              height="41"
              src="img/logo.svg"
              width="81"
            />
          </a>
        </div>
        <UserNav />
      </div>
    </div>
  </header>
  <main
    className="page__main page__main--property"
  >
    <section
      className="property"
    >
      <div
        className="property__gallery-container container"
      >
        <div
          className="property__gallery"
        />
      </div>
      <div
        className="property__container container"
      >
        <div
          className="property__wrapper"
        >
          <div
            className="property__mark"
          >
            <span>
              Premium
            </span>
          </div>
          <div
            className="property__name-wrapper"
          >
            <h1
              className="property__name"
            >
              
            </h1>
            <button
              className="property__bookmark-button button property__bookmark-button--active"
              onClick={[Function]}
              type="button"
            >
              <svg
                className="property__bookmark-icon"
                height={33}
                width={31}
              >
                <use
                  xlinkHref="#icon-bookmark"
                />
              </svg>
              <span
                className="visually-hidden"
              >
                To bookmarks
              </span>
            </button>
          </div>
          <div
            className="property__rating rating"
          >
            <div
              className="property__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
              <span
                className="visually-hidden"
              >
                Rating
              </span>
            </div>
            <span
              className="property__rating-value rating__value"
            >
              0
            </span>
          </div>
          <ul
            className="property__features"
          >
            <li
              className="property__feature property__feature--entire"
            >
              
            </li>
            <li
              className="property__feature property__feature--bedrooms"
            >
              0
               Bedrooms
            </li>
            <li
              className="property__feature property__feature--adults"
            >
              Max 
              0
               adults
            </li>
          </ul>
          <div
            className="property__price"
          >
            <b
              className="property__price-value"
            >
              €
              0
            </b>
            <span
              className="property__price-text"
            >
               night
            </span>
          </div>
          <div
            className="property__inside"
          >
            <h2
              className="property__inside-title"
            >
              What's inside
            </h2>
            <ul
              className="property__inside-list"
            />
          </div>
          <div
            className="property__host"
          >
            <h2
              className="property__host-title"
            >
              Meet the host
            </h2>
            <div
              className="property__host-user user"
            >
              <div
                className="property__avatar-wrapper user__avatar-wrapper property__avatar-wrapper--pro"
              >
                <img
                  alt="Host avatar"
                  className="property__avatar user__avatar"
                  height="74"
                  src=""
                  width="74"
                />
              </div>
              <span
                className="property__user-name"
              >
                
              </span>
            </div>
            <div
              className="property__description"
            >
              <p
                className="property__text"
              >
                Hello, mello, shmello
              </p>
            </div>
          </div>
          <section
            className="property__reviews reviews"
          >
            <h2
              className="reviews__title"
            >
              Reviews · 
              <span
                className="reviews__amount"
              >
                1
              </span>
            </h2>
            <ul
              className="reviews__list"
            >
              <li
                className="reviews__item"
              >
                <div
                  className="reviews__user user"
                >
                  <div
                    className="reviews__avatar-wrapper user__avatar-wrapper"
                  >
                    <img
                      alt="Reviews avatar"
                      className="reviews__avatar user__avatar"
                      height="54"
                      src=""
                      width="54"
                    />
                  </div>
                  <span
                    className="reviews__user-name"
                  >
                    
                  </span>
                </div>
                <div
                  className="reviews__info"
                >
                  <div
                    className="reviews__rating rating"
                  >
                    <div
                      className="reviews__stars rating__stars"
                    >
                      <span
                        style={
                          Object {
                            "width": "100%",
                          }
                        }
                      />
                      <span
                        className="visually-hidden"
                      >
                        Rating
                      </span>
                    </div>
                  </div>
                  <p
                    className="reviews__text"
                  >
                    
                  </p>
                  <time
                    className="reviews__time"
                    dateTime="2020-12-02T17:54:54.499Z"
                  >
                    December 2020
                  </time>
                </div>
              </li>
            </ul>
            <CommentForm
              id={0}
            />
          </section>
        </div>
      </div>
      <section
        className="property__map map"
      >
        <CityMap
          activeOffer={
            Object {
              "adults": 0,
              "bedrooms": 0,
              "city": Object {
                "name": "",
              },
              "cost": 0,
              "description": "Hello, mello, shmello",
              "goods": Array [],
              "host": Object {
                "avatarUrl": "",
                "id": 0,
                "isPro": true,
                "name": "",
              },
              "id": 0,
              "images": Array [],
              "isFavorite": true,
              "isPremium": true,
              "previewImg": "",
              "rate": 0,
              "title": "",
              "type": "",
            }
          }
          offers={
            Array [
              Object {
                "adults": 0,
                "bedrooms": 0,
                "city": Object {
                  "name": "",
                },
                "cost": 0,
                "description": "Hello, mello, shmello",
                "goods": Array [],
                "host": Object {
                  "avatarUrl": "",
                  "id": 0,
                  "isPro": true,
                  "name": "",
                },
                "id": 0,
                "images": Array [],
                "isFavorite": true,
                "isPremium": true,
                "previewImg": "",
                "rate": 0,
                "title": "",
                "type": "",
              },
            ]
          }
          styles={
            Object {
              "marginLeft": "auto",
              "marginRight": "auto",
              "width": "1144px",
            }
          }
          type="BIG"
        />
      </section>
    </section>
    <div
      className="container"
    >
      <section
        className="near-places places"
      >
        <h2
          className="near-places__title"
        >
          Other places in the neighbourhood
        </h2>
        <div
          className="near-places__list places__list"
        >
          <article
            className="near-places__card place-card"
            onMouseOver={[Function]}
          >
            <div
              className="place-card__mark"
            >
              <span>
                Premium
              </span>
            </div>
            <div
              className="near-places__image-wrapper place-card__image-wrapper"
            >
              <a
                href="/offer/0"
                onClick={[Function]}
              >
                <img
                  className="place-card__image"
                  height="200"
                  src=""
                  width="260"
                />
              </a>
            </div>
            <div
              className="place-card__info"
            >
              <div
                className="place-card__price-wrapper"
              >
                <div
                  className="place-card__price"
                >
                  <b
                    className="place-card__price-value"
                  >
                    €
                    0
                  </b>
                  <span
                    className="place-card__price-text"
                  >
                    / night
                  </span>
                </div>
                <button
                  className="place-card__bookmark-button button place-card__bookmark-button--active"
                  onClick={[Function]}
                  type="button"
                >
                  <svg
                    className="place-card__bookmark-icon"
                    height={19}
                    width={18}
                  >
                    <use
                      xlinkHref="#icon-bookmark"
                    />
                  </svg>
                  <span
                    className="visually-hidden"
                  >
                    To bookmarks
                  </span>
                </button>
              </div>
              <div
                className="place-card__rating rating"
              >
                <div
                  className="place-card__stars rating__stars"
                >
                  <span
                    style={
                      Object {
                        "width": "0%",
                      }
                    }
                  />
                  <span
                    className="visually-hidden"
                  >
                    Rating
                  </span>
                </div>
              </div>
              <h2
                className="place-card__name"
              >
                <a
                  href="/offer/0"
                  onClick={[Function]}
                >
                  
                </a>
              </h2>
              <p
                className="place-card__type"
              >
                
              </p>
            </div>
          </article>
        </div>
      </section>
    </div>
  </main>
</div>
`;
